### PR TITLE
Added support for meta in bwamem2/index

### LINKED ISSUE #1917: bwamem2/index and minimap2/index don't consider meta
### Have you checked the docs?

- [X] [nf-core website: troubleshooting](https://nf-co.re/usage/troubleshooting)
- [X] [nf-core modules documentation](https://nf-co.re/docs/contributing/modules)

### Description of the bug

I've found out that bwamem2/index and minimap2/index don't consider meta in their input and output. Is there a reason why ?
Otherwise, I'm happy to go update the modules.

### Command used and terminal output

_No response_

### Relevant files

https://github.com/nf-core/modules/blob/a8b0fce8ce1f44ff3ee776639431a2472f9f308b/modules/bwamem2/index/main.nf#L11
https://github.com/nf-core/modules/blob/a8b0fce8ce1f44ff3ee776639431a2472f9f308b/modules/minimap2/index/main.nf#L10

### System information

_No response_

### PR DESCRIPTION
**Pull request by muffato**
I think bwamem2/index and minimap2/index should consider meta in their input and output. This would make it much easier to plug them in workflows (no need to do `.map { it[1] }`).

## PR checklist

Closes #1917

- [x] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by muffato** [comment]
Of course I need to update the tests ! 🤦🏼

**Comment by muffato** [comment]
My understanding of meta is that it's a way of propagating additional information (metadata) throughout a workflow. Say the workflow is 1) get a genome, 2) index the genome, 3) align some reads; having to strip out meta before indexing, and then put it back for the alignment (`bwamem2/align` does expect meta) is convoluted and awkward. Why can't `bwamem2/index` use meta in the first place ?

**Comment by nvnieuwk** [comment]
`bwamem2/mem` (which is what I think you mean with `bwamem2/align` does not require meta for the index (https://github.com/nf-core/modules/blob/master/modules/bwamem2/mem/main.nf)). It would be a bit weird to add the meta to this IMO since it's a main file used for all samples.

**Comment by nvnieuwk** [comment]
@jfy133 what is your opinion on this?

**Comment by jfy133** [comment]
> Why would you need to use meta for a mapper index? Isn't it the same for all samples of the same species, so you just have to use one? (Just trying to understand your reasoning :p feel free to enlighten me :) )

What if your pipeline needs to map against multiple species/references!

I'm actually all for this. I'm more and more just saying add `meta` to everything, because it costs 'nothing' to add an empty input for meta.

But I think such a change will require a large update across all aligners, and update to the guidelines. Lets bring it up there

**Comment by nvnieuwk** [comment]
That indeed sounds very reasonable now. I haven't thought of it that way!

**Comment by jfy133** [comment]
> That indeed sounds very reasonable now. I haven't thought of it that way!

Genomicists! :shakesfist: You sound like those 'sarek' people ;) (affectionate teasing pinging @maxulysse @FriederikeHanssen)

**Comment by muffato** [comment]
@nvnieuwk . Ah, right ! In my comment, I indeed got confused between the meta map associated with the reference genome, and the meta map associated with the reads. You're right: `bwamem2/mem` only accepts the latter, not the former.
Base on your comment alone, as a compromise, I would have suggested to make `bwamem2/index` take the meta map of the fasta file, but discard it and output the index alone. My motivation for this pull-request was to be able to chain modules directly without having to edit the inputs/outputs, so the easiest would have been to keep the index free of meta in both `index` and `mem`.
But if the consensus is that the index should keep a meta map, then indeed, it should be added to `mem` too. But `mem` would then take two meta maps. Should it merge them for its output (named `bam`) ?

**Comment by nvnieuwk** [comment]
> 

No problem about the confusion, it happens to us all :p. I think this could open up a very interesting discussion, let's see where it will lead!

**Comment by jfy133** [comment]
> But if the consensus is that the index should keep a meta map, then indeed, it should be added to `mem` too. But `mem` would then take two meta maps. Should it merge them for its output (named `bam`) ?

That is indeed the one tricky thing here... and probably the reason why it wasn't directl implemented before (correct me if I'm wrong @drpatelh )? 

You could have clashes e.g. two `ids` in which you wouldn't necessarily know what to select (should you use `meta.id` or `meta2.id` for example) . On the flipside, the pipeline developer could modify this as necessary between modules, and if you don't want any meta - a blank `[]` is still fine.

**Comment by drpatelh** [comment]
I think everyone is right here for one reason or another! 😅 

The reason it wasn't added inititally is because the `meta` map was reserved for sample-level metadata (and not generic metadata). If your pipeline uses one genome for all samples then the indexing modules don't need any sample-level metadata. However, if your pipeline has been written to accept separate reference genomes per sample e.g. [nanoseq](https://nf-co.re/nanoseq/3.0.0/usage#introduction) then it makes sense to have the reference genome info in the `meta` map too so you can propagate that through the pipeline as required.

Overall, I think we should add it in for all indexing modules because the only downside if it isn't defined is that you need to do a little channel manipulation or provide empty `meta` maps. However, if you do required the usage of one genome per sample then you can't use the modules in their current state.

**Comment by maxulysse** [comment]
Agree, we should add `meta` to all modules.

**Comment by muffato** [comment]
Hi. Here is how it would look like for bwamem2. `mem` takes two meta maps, and merges them, the reads' map overriding the index's in case the same key is used in both.

**Comment by drpatelh** [comment]
Bit confused as to why we need 2 `meta` maps if the `meta` map is always required for sample metadata only.

**Comment by maxulysse** [comment]
> Hi. Here is how it would look like for bwamem2. `mem` takes two meta maps, and merges them, the reads' map overriding the index's in case the same key is used in both.

Same here

**Comment by maxulysse** [comment]
Oooh I see, you want to give out the reference with a meta map as well...
Why would you want to do so?
Can you elaborate a use case for that?

**Comment by muffato** [comment]
> Bit confused as to why we need 2 `meta` maps if the `meta` map is always required for sample metadata only.

> Why would you want to do so?
> Can you elaborate a use case for that?

I thought that's what decided with https://github.com/nf-core/modules/pull/1921#issuecomment-1193947929 and https://github.com/nf-core/modules/pull/1921#issuecomment-1194106070 😕

**Comment by drpatelh** [comment]
Let's keep the current standard notation:

```
input:
tuple val(meta), path(fasta)
```

File: modules/bwamem2/mem/main.nf [hunk 1]
```diff
@@ -9,17 +9,18 @@ process BWAMEM2_MEM {
...
-    tuple val(meta), path("*.bam"), emit: bam
-    path  "versions.yml"          , emit: versions
+    tuple val(merged_meta), path("*.bam"), emit: bam
+    path  "versions.yml"                 , emit: versions
 
     when:
     task.ext.when == null || task.ext.when
 
     script:
+    merged_meta = meta2 + meta    // Priority given to the reads' meta map
```
Comment (drpatelh) [inline modules/bwamem2/mem/main.nf:23]:
```suggestion
```

File: modules/bwamem2/mem/main.nf [hunk 2]
```diff
@@ -9,17 +9,18 @@ process BWAMEM2_MEM {
...
     tuple val(meta), path(reads)
-    path  index
+    tuple val(meta2), path(index)
     val   sort_bam
 
     output:
-    tuple val(meta), path("*.bam"), emit: bam
-    path  "versions.yml"          , emit: versions
+    tuple val(merged_meta), path("*.bam"), emit: bam
+    path  "versions.yml"                 , emit: versions
```
Comment (drpatelh) [inline modules/bwamem2/mem/main.nf:17]:
```suggestion
    tuple val(meta), path("*.bam"), emit: bam
    path  "versions.yml"          , emit: versions
```

File: modules/bwamem2/mem/meta.yml [hunk 3]
```diff
@@ -21,13 +21,18 @@ input:
...
   - reads:
       type: file
       description: |
         List of input FastQ files of size 1 and 2 for single-end and paired-end data,
         respectively.
+  - meta2:
+      type: map
+      description: |
+        Groovy Map containing sample information (about the reference genome)
+        e.g. [ id:'mus_musculus', assembly:'GRCm38' ]
```
Comment (drpatelh) [inline modules/bwamem2/mem/meta.yml:35]:
```suggestion
```

File: modules/bwamem2/mem/meta.yml [hunk 4]
```diff
@@ -21,13 +21,18 @@ input:
   - meta:
       type: map
       description: |
-        Groovy Map containing sample information
+        Groovy Map containing sample information (about the reads)
```
Comment (drpatelh) [inline modules/bwamem2/mem/meta.yml:24]:
```suggestion
        Groovy Map containing sample information
```

File: modules/bwamem2/mem/meta.yml [hunk 5]
```diff
@@ -37,6 +42,11 @@ input:
       description: use samtools sort (true) or samtools view (false)
       pattern: "true or false"
 output:
+  - merged_meta:
+      type: map
+      description: |
+        Groovy Map containing sample information, combined from the reads and index parameters
+        e.g. [ id:'test', single_end:false, assembly:'GRCm38' ]
```
Comment (drpatelh) [inline modules/bwamem2/mem/meta.yml:49]:
```suggestion
  - meta:
      type: map
      description: |
        Groovy Map containing sample information
        e.g. [ id:'test', single_end:false ]
```

**Review by drpatelh — COMMENTED** [review]
🚀

**Review by drpatelh — APPROVED** [review]
Thanks @muffato !

**Comment by muffato** [comment]
Thanks all for your patience, and sorry I misunderstood the explanations

**Pull request merged**